# Prétraitement et réglage fin du transformateur....
Dernière mise à jour : 17 juillet 2025

Défi quotidien : prétraiter et affiner les modèles basés sur des transformateurs


👩‍🏫 👩🏿‍🏫 Ce que vous apprendrez
Dans ce défi quotidien, vous apprendrez à prétraiter et à affiner les modèles basés sur des transformateurs, notamment BERT et XLM-RoBERTa, pour les tâches de classification de texte. Vous comprendrez :

Comment fonctionne la tokenisation pour ces modèles.
Comment formater correctement les données d'entrée.
Comment affiner les modèles de transformateurs pour les tâches de classification.
Comment effectuer une validation croisée à l'aide du fractionnement k-fold.


🛠️ Ce que vous allez créer
À la fin de ce défi, vous disposerez d'un modèle de transformateur optimisé (BERT ou XLM-RoBERTa) capable de classer du texte en différentes catégories. De plus, vous structurerez les données pour l'entraînement, les validerez par validation croisée et comprendrez comment optimiser ces modèles pour de meilleures performances.



Ensemble de données
Vous pouvez trouver l'ensemble de données pour cet exercice ici



Tâche


1. Comprendre BERT et XLM-RoBERTa
Objectif : Apprendre comment fonctionnent les modèles de transformateurs et leur rôle dans les tâches de PNL.

Instructions:

Lisez les descriptions de BERT et XLM-RoBERTa.
Comprendre comment ces modèles traitent le texte à l’aide de la tokenisation.
Découvrez les différentes versions pré-entraînées de ces modèles et leurs caractéristiques.
Fonctions à utiliser :

from transformers import BertTokenizer, XLMRobertaTokenizer


2. Tokenisation du texte
Objectif : comprendre comment tokeniser du texte à l’aide de tokeniseurs pré-entraînés.

Instructions:

Utilisez BertTokenizeret XLMRobertaTokenizerpour convertir des phrases en entrées tokenisées.
Explorez les différents types de jetons, tels que input_ids, attention_mask, et labels.
Expérimentez la tokenisation d’une seule phrase et de deux phrases.
Fonctions à utiliser :

tokenizer.encode_plus()
tokenizer.decode()


3. Préparation des données d'entrée pour le modèle
Objectif : Formater correctement les données d'entrée pour les modèles de transformateurs.

Instructions:

Assurez-vous que les phrases d’entrée sont complétées ou tronquées à la longueur souhaitée.
Comprendre et définir des jetons spéciaux tels que <s>et </s>.
Découvrez attention_maskcomment cela aide le modèle à ignorer les jetons de remplissage.
Fonctions à utiliser :

tokenizer.encode_plus()
tokenizer.special_tokens_map
tokenizer.vocab_size


4. Chargement et exploration de l'ensemble de données
Objectif : Charger l’ensemble de données et explorer sa structure.

Instructions:

Chargez les données de formation et de test à partir de fichiers CSV.
Affichez les premières lignes pour comprendre sa structure.
Identifiez les colonnes nécessaires à la formation du modèle.
Fonctions à utiliser :

pd.read_csv()
df.head()
df.shape


5. Création de plis de validation croisée
Objectif : Mettre en œuvre une validation croisée k-fold pour la formation.

Instructions:

Permet StratifiedKFoldde créer 5 divisions de validation de formation.
Assurez-vous que chaque pli conserve la même répartition des étiquettes.
Stockez les divisions de formation et de validation dans des listes séparées.
Fonctions à utiliser :

from sklearn.model_selection import StratifiedKFold
kf.split()
shuffle()


## 1. Comprendre BERT et XLM-RoBERTa

In [1]:
# ==============================================
# INSTALLATION DES DEPENDANCES
# ==============================================
!pip install --quiet transformers sentencepiece

# ==============================================
# IMPORTATION DES LIBRAIRIES
# ==============================================
from transformers import BertTokenizer, XLMRobertaTokenizer

# ==============================================
# INITIALISATION DES TOKENIZERS
# ==============================================
# BERT : modèle anglais
bert_tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")

# XLM-RoBERTa : modèle multilingue (100 langues)
xlm_tokenizer = XLMRobertaTokenizer.from_pretrained("xlm-roberta-base")

# ==============================================
# TOKENISATION DE TEXTES D'EXEMPLE
# ==============================================
texte_anglais = "Hello, how are you?"
tokens_bert = bert_tokenizer.tokenize(texte_anglais)

texte_francais = "Bonjour, comment ça va ?"
tokens_xlm = xlm_tokenizer.tokenize(texte_francais)

# ==============================================
# AFFICHAGE
# ==============================================
print("Tokenisation BERT :", tokens_bert)
print("Tokenisation XLM-RoBERTa :", tokens_xlm)

# ==============================================
# EXPLICATION :
# BERT : WordPiece (préfixe '##'), optimisé anglais.
# XLM-RoBERTa : BPE (préfixe '▁'), multilingue (100 langues).


c:\Users\chume\.conda\envs\llama\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
c:\Users\chume\.conda\envs\llama\lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\chume\.cache\huggingface\hub\models--xlm-roberta-base. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, 

Tokenisation BERT : ['hello', ',', 'how', 'are', 'you', '?']
Tokenisation XLM-RoBERTa : ['▁Bonjour', ',', '▁comment', '▁ça', '▁va', '▁?']


Analyse rapide :

* **Résultats de tokenisation** : corrects.

  * BERT découpe mot à mot sans sous-mots car la phrase est simple.
  * XLM-RoBERTa utilise les préfixes `▁`, typique du BPE multilingue, normal.

* **Avertissements système** :

  * **tqdm/IProgress** : widgets Jupyter non à jour → solution rapide :

    ```bash
    pip install --upgrade ipywidgets
    jupyter nbextension enable --py widgetsnbextension
    ```
  * **huggingface\_hub symlinks** : Windows interdit les liens symboliques sans mode développeur. ⚠️ Impact : cache moins efficace, pas de blocage de fonctionnement.
  * **FutureWarning resume\_download** : prévient d’un futur changement d’option, **pas d’impact actuel**.

**Conclusion** : ton code fonctionne parfaitement, ces messages sont juste informatifs.
Tu peux les ignorer ou corriger via mise à jour de `ipywidgets` et activation du mode développeur Windows.


Voici exactement ce que tu as dû faire pour faire fonctionner le code lié à BERT et XLM-RoBERTa :

---

### Étapes nécessaires pour faire fonctionner le code :

1. **Créer un environnement propre (éviter l’environnement `base`)** :

   ```bash
   conda create -n monenv python=3.10 -c conda-forge
   ```

2. **Activer cet environnement** :

   ```bash
   conda activate monenv
   ```

3. **Installer les dépendances dans cet environnement** :

   ```bash
   conda install -c conda-forge transformers sentencepiece
   ```

4. **Installer Jupyter Notebook si nécessaire** :

   ```bash
   pip install notebook
   ```

5. **Lancer Jupyter Notebook depuis le même environnement** :

   ```bash
   jupyter notebook
   ```

---

### 🔧 Pourquoi c’était nécessaire :

* L’environnement `base` ne donnait **pas les droits d’écriture** pour installer `sentencepiece`.
* `XLMRobertaTokenizer` **dépend obligatoirement** de `sentencepiece` (non inclus par défaut avec `transformers`).
* Un environnement conda isolé (`monenv`) te donne **tous les droits** et évite les conflits de version.

---

Résultat : tu as pu exécuter le code sans erreur avec les deux tokenizers (`BertTokenizer` et `XLMRobertaTokenizer`).


## 2. Tokenisation du texte

In [2]:
# ==============================================
# INSTALLATION DES DEPENDANCES
# ==============================================
!pip install --quiet transformers sentencepiece

# ==============================================
# IMPORTATION DES LIBRAIRIES
# ==============================================
from transformers import BertTokenizer, XLMRobertaTokenizer

# ==============================================
# INITIALISATION DES TOKENIZERS
# ==============================================
bert_tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
xlm_tokenizer = XLMRobertaTokenizer.from_pretrained("xlm-roberta-base")

# ==============================================
# EXEMPLE : UNE PHRASE (BERT)
# ==============================================
texte = "Hello, how are you?"
encoding_bert = bert_tokenizer.encode_plus(
    texte,
    truncation=True,
    padding='max_length',
    max_length=10,
    return_tensors='pt'
)

print("BERT input_ids :", encoding_bert['input_ids'])
print("BERT attention_mask :", encoding_bert['attention_mask'])
print("Décodage BERT :", bert_tokenizer.decode(encoding_bert['input_ids'][0]))

# ==============================================
# EXEMPLE : DEUX PHRASES (XLM-RoBERTa)
# ==============================================
texte1 = "Bonjour, comment ça va ?"
texte2 = "Très bien merci."

encoding_xlm = xlm_tokenizer.encode_plus(
    texte1,
    texte2,
    truncation=True,
    padding='max_length',
    max_length=15,
    return_tensors='pt'
)

print("XLM input_ids :", encoding_xlm['input_ids'])
print("XLM attention_mask :", encoding_xlm['attention_mask'])
print("Décodage XLM :", xlm_tokenizer.decode(encoding_xlm['input_ids'][0]))

# ==============================================
# EXPLICATIONS RAPIDES :
# ==============================================
# input_ids : ID numérique de chaque token.
# attention_mask : 1 = token utile, 0 = padding.
# labels : utilisés uniquement pour l'entraînement (pas montré ici).
# encode_plus() : produit tous ces éléments en une seule étape.
# decode() : permet de revenir des ID aux mots originaux.


BERT input_ids : tensor([[ 101, 7592, 1010, 2129, 2024, 2017, 1029,  102,    0,    0]])
BERT attention_mask : tensor([[1, 1, 1, 1, 1, 1, 1, 1, 0, 0]])
Décodage BERT : [CLS] hello, how are you? [SEP] [PAD] [PAD]
XLM input_ids : tensor([[     0,  84602,      4,   6868,   3244,    307,    705,      2,      2,
         103132,   1806,  38741,      5,      2,      1]])
XLM attention_mask : tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0]])
Décodage XLM : <s> Bonjour, comment ça va?</s></s> Très bien merci.</s><pad>


Analyse rapide :

###  **BERT**

* `input_ids` : OK → ajout des tokens spéciaux `[CLS]` (101), `[SEP]` (102), et padding (0).
* `attention_mask` : 1 = vrais tokens, 0 = padding (correct).
* `decode()` : bonne reconstruction, **les `[PAD]` sont visibles**, mais non pris en compte par `attention_mask` → normal.

###  **XLM-RoBERTa**

* `input_ids` : OK → `<s>` (0), `</s>` (2), `<pad>` (1), avec séparation correcte des phrases.
* `attention_mask` : bonne gestion du padding (1 pour tokens valides, 0 pour padding).
* `decode()` : les balises `<s>` et `</s>` sont visibles, **normal**, car XLM-RoBERTa conserve explicitement ces tokens.

---

###  **Conclusion** :

* Les **tokens spéciaux sont bien intégrés**.
* L'**attention\_mask est correct**, prêt pour entraînement.
* **Décodage fonctionne** malgré affichage visible des tokens spéciaux — normal avec `decode()` sans nettoyage supplémentaire.

**Tout est fonctionnel**.


## 3. Préparation des données d'entrée pour le modèle

In [3]:
# ==============================================
# INSTALLATION DES DEPENDANCES
# ==============================================
!pip install --quiet transformers sentencepiece

# ==============================================
# IMPORTATION
# ==============================================
from transformers import BertTokenizer, XLMRobertaTokenizer

# ==============================================
# INITIALISATION
# ==============================================
bert_tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
xlm_tokenizer = XLMRobertaTokenizer.from_pretrained("xlm-roberta-base")

# ==============================================
# TOKENISATION AVEC PADDING ET TRONQUAGE
# ==============================================
texte = "Ceci est un exemple de phrase pour la démonstration."

encoding = xlm_tokenizer.encode_plus(
    texte,
    truncation=True,              # coupe si trop long
    padding='max_length',         # complète jusqu'à max_length
    max_length=12,                # longueur cible
    return_attention_mask=True,
    return_tensors='pt'           # tensors PyTorch
)

print("input_ids :", encoding['input_ids'])
print("attention_mask :", encoding['attention_mask'])

# ==============================================
# JETONS SPÉCIAUX
# ==============================================
print("Jetons spéciaux (XLM-RoBERTa) :", xlm_tokenizer.special_tokens_map)

# Exemples typiques :
# <s> = début de séquence
# </s> = fin de séquence
# <pad> = remplissage (padding)
# <unk> = mot inconnu
# <mask> = pour le masquage lors de l'entraînement

# ==============================================
# TAILLE DU VOCABULAIRE
# ==============================================
print("Taille du vocabulaire XLM-RoBERTa :", xlm_tokenizer.vocab_size)

# ==============================================
# EXPLICATION RAPIDE
# ==============================================
# - input_ids = tokens convertis en IDs numériques
# - attention_mask = 1 pour tokens réels, 0 pour padding
# - special_tokens_map = dictionnaire des jetons spéciaux
# - vocab_size = taille totale du vocabulaire connu par le modèle


input_ids : tensor([[     0, 124532,    437,     51,  18632,      8, 113860,    578,     21,
          91744,   2816,      2]])
attention_mask : tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])
Jetons spéciaux (XLM-RoBERTa) : {'bos_token': '<s>', 'eos_token': '</s>', 'unk_token': '<unk>', 'sep_token': '</s>', 'pad_token': '<pad>', 'cls_token': '<s>', 'mask_token': '<mask>'}
Taille du vocabulaire XLM-RoBERTa : 250002


Analyse rapide :

* **input\_ids** : séquence encodée avec `<s>` (0) au début et `</s>` (2) à la fin → **normal** pour XLM-RoBERTa.

* **attention\_mask** : uniquement des 1 → aucune troncature ni padding (12 tokens, donc inférieur au `max_length` probable).

* **special\_tokens\_map** : cohérent pour XLM-RoBERTa :

    * `<s>` et `</s>` servent de début/fin.
    * `<pad>` pour le padding.
    * `<mask>` pour le masquage MLM.
    * `<unk>` pour tokens inconnus.

* **vocab\_size** : 250002, valeur correcte pour `xlm-roberta-base`.

* **Conclusion** :
Tout est **conforme**, séquences prêtes pour le modèle, vocabulaire complet et tokenisation bien encadrée par les tokens spéciaux.


## 4. Chargement et exploration de l'ensemble de données

In [4]:
# ==============================================
# INSTALLATION DES LIBRAIRIES
# ==============================================
!pip install --quiet pandas

# ==============================================
# IMPORTATION
# ==============================================
import pandas as pd

# ==============================================
# CHARGEMENT D'UN FICHIER CSV
# ==============================================
# Remplace 'train.csv' et 'test.csv' par le chemin réel
train_df = pd.read_csv("train.csv")
test_df = pd.read_csv("test.csv")

# ==============================================
# EXPLORATION RAPIDE
# ==============================================
print("Train - Premières lignes :")
print(train_df.head())

print("\nTest - Premières lignes :")
print(test_df.head())

print("\nTaille du jeu d'entraînement :", train_df.shape)
print("Taille du jeu de test :", test_df.shape)

# ==============================================
# ANALYSE DE LA STRUCTURE
# ==============================================
print("\nColonnes disponibles :", train_df.columns.tolist())

# ==============================================
# EXPLICATION :
# - head() = aperçu rapide des données
# - shape = nombre de lignes et colonnes
# - columns = noms de colonnes pour identifier texte + labels


Train - Premières lignes :
           id                                            premise  \
0  5130fd2cb5  and these comments were considered in formulat...   
1  5b72532a0b  These are issues that we wrestle with in pract...   
2  3931fbe82a  Des petites choses comme celles-là font une di...   
3  5622f0c60b  you know they can't really defend themselves l...   
4  86aaa48b45  ในการเล่นบทบาทสมมุติก็เช่นกัน โอกาสที่จะได้แสด...   

                                          hypothesis lang_abv language  label  
0  The rules developed in the interim were put to...       en  English      0  
1  Practice groups are not permitted to work on t...       en  English      2  
2              J'essayais d'accomplir quelque chose.       fr   French      0  
3  They can't defend themselves because of their ...       en  English      0  
4    เด็กสามารถเห็นได้ว่าชาติพันธุ์แตกต่างกันอย่างไร       th     Thai      1  

Test - Premières lignes :
           id                                            

Analyse rapide :

 **Structure correcte** :

* **train\_df** : 12120 lignes, 6 colonnes → données pour classification.
* **test\_df** : 5195 lignes, pas de colonne `label` → normal pour test sans supervision.

 **Colonnes clés** :

* `premise`, `hypothesis` : colonnes de texte à concaténer/tokeniser.
* `label` (train seulement) : cible (classification).
* `lang_abv`, `language` : info langue → utile pour multilingue.

 **Observation** :

* Données multilingues (fr, en, th, ar, ur, zh, ru, etc.), idéal pour **XLM-RoBERTa**.
* `id` = identifiant unique.

 **Conclusion** : dataset bien structuré, il faudra utiliser `premise` + `hypothesis` comme entrées, `label` comme sortie.
**Train** : texte + label
**Test** : texte uniquement.


## 5. Création de plis de validation croisée

In [5]:
# ==============================================
# INSTALLATION DES LIBRAIRIES
# ==============================================
!pip install --quiet pandas scikit-learn

# ==============================================
# IMPORTATIONS
# ==============================================
import pandas as pd
from sklearn.model_selection import StratifiedKFold

# ==============================================
# CHARGEMENT D'UN EXEMPLE DE DATASET
# ==============================================
# Exemple rapide : dataframe fictif
data = {'text': [f"phrase {i}" for i in range(20)],
        'label': [0, 1]*10}
df = pd.DataFrame(data)

print("Structure du dataset :", df.shape)
print(df.head())

# ==============================================
# STRATIFIED KFOLD CROSS VALIDATION
# ==============================================
kf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

train_indices = []
val_indices = []

for fold, (train_idx, val_idx) in enumerate(kf.split(df, df['label'])):
    print(f"\nPli {fold+1}:")
    print("Train indices :", train_idx)
    print("Validation indices :", val_idx)
    
    train_indices.append(train_idx)
    val_indices.append(val_idx)

# ==============================================
# EXPLICATION :
# - StratifiedKFold : équilibre des étiquettes dans chaque pli.
# - shuffle=True : mélange avant découpage.
# - train_indices et val_indices : contiennent les indices pour l'entraînement et la validation.


Structure du dataset : (20, 2)
       text  label
0  phrase 0      0
1  phrase 1      1
2  phrase 2      0
3  phrase 3      1
4  phrase 4      0

Pli 1:
Train indices : [ 0  4  5  7  8  9 10 11 12 13 14 15 16 17 18 19]
Validation indices : [1 2 3 6]

Pli 2:
Train indices : [ 0  1  2  3  4  5  6  7  8 11 12 13 14 15 17 18]
Validation indices : [ 9 10 16 19]

Pli 3:
Train indices : [ 0  1  2  3  5  6  8  9 10 12 13 15 16 17 18 19]
Validation indices : [ 4  7 11 14]

Pli 4:
Train indices : [ 0  1  2  3  4  5  6  7  9 10 11 12 14 15 16 19]
Validation indices : [ 8 13 17 18]

Pli 5:
Train indices : [ 1  2  3  4  6  7  8  9 10 11 13 14 16 17 18 19]
Validation indices : [ 0  5 12 15]


Analyse rapide :

 **Dataset** :

* 20 lignes, 2 colonnes (`text`, `label`), échantillon de test pour démonstration.

 **Validation croisée stratifiée (5 plis)** :

* Chaque pli contient **4 indices validation** (20 ÷ 5 = 4), et 16 indices entraînement.
* La séparation est correcte : les indices validation changent à chaque pli.
* Comme les labels sont équilibrés (0,1 alternés), **StratifiedKFold** a bien respecté la proportion des classes dans chaque pli.

 **Conclusion** :
StratifiedKFold fonctionne correctement : bonne répartition, pas de fuite de données, préparation idéale pour entraînement + validation fiables.


## 6. Bilan de cet exercice

 **Ce que tu as fait** :

* Compris **BERT** et **XLM-RoBERTa** (fonctionnement, tokenisation, différences).
* Appliqué **tokenisation avancée** avec `encode_plus()`, `decode()`, `attention_mask`, et jetons spéciaux.
* Préparé les **données d’entrée correctement** : padding, troncature, gestion des tokens spéciaux.
* **Exploré un dataset multilingue**, identifié `premise`, `hypothesis`, `label`.
* Mis en place une **validation croisée stratifiée** fiable (StratifiedKFold).

 **Compétences validées** :

* Maitrise des tokenizers Huggingface.
* Préparation propre des données pour modèles Transformers.
* Manipulation basique de datasets multilingues.
* Utilisation correcte de la validation croisée supervisée.

 **Conclusion** : maîtrise fonctionnelle de la **préparation complète** d’un pipeline NLP moderne avec Transformers.

Prêt pour le fine-tuning et la création de modèles complets.


## 7. Applications innovantes et creatives

Voici **5 idées de projets innovants et concrets** basés sur ce que tu as pratiqué:

---

### 1. **Détecteur multilingue de contradictions**

**Objectif** : entraîner un modèle XLM-RoBERTa sur un dataset type **Natural Language Inference (NLI)** pour détecter si deux phrases sont cohérentes, contradictoires ou neutres.

*  Bonus : Interface Streamlit multilingue.
* 📈 Utilité : utile pour la détection automatique de fake news, vérification d’informations.

---

### 2. **Résumé automatique multilingue avec BERTSum ou RoBERTa**

**Objectif** : utiliser BERT ou XLM-RoBERTa pour générer des résumés courts de textes en plusieurs langues.

*  Bonus : comparer les performances entre langues peu représentées et riches.
*  Ajout créatif : possibilité de choisir un “style” de résumé (formel, familier…).

---

### 3. **Chatbot multilingue à base de NLI**

**Objectif** : créer un **chatbot intelligent** qui comprend une demande en plusieurs langues et choisit une réponse logique parmi un ensemble pré-défini (classification par hypothèses).

*  Bonus : intégration rapide avec FastAPI.
*  Application : service client multilingue.

---

### 4. **Classification de sentiments multilingue**

**Objectif** : créer un classificateur de sentiments (positif, négatif, neutre) sur des données sociales multilingues avec XLM-RoBERTa.

*  Bonus : adaptation sur tweets, reviews Amazon, Reddit.
*  Extension : dashboard avec visualisations des émotions par langue.

---

### 5. **Filtre de contenu toxique multilingue**

**Objectif** : fine-tuning d’un modèle RoBERTa pour détecter automatiquement les propos haineux ou offensants dans plusieurs langues.

*  Bonus : possibilité de filtrer selon la gravité du propos.
*  Utilité : modération automatique pour forums/sites.

---
